server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -730,40 +729,19 @@ void executeBulk(
     private static class ConcreteIndices {
         private final ClusterState state;
         private final IndexNameExpressionResolver indexNameExpressionResolver;
-        private final Map<String, Index> indices = new HashMap<>();
+        private final Map<String, IndexAbstraction> indexAbstractions = new HashMap<>();
         private final Map<Index, IndexRouting> routings = new HashMap<>();
 
         ConcreteIndices(ClusterState state, IndexNameExpressionResolver indexNameExpressionResolver) {
             this.state = state;
             this.indexNameExpressionResolver = indexNameExpressionResolver;
         }
 
-        Index getConcreteIndex(String indexOrAlias) {
-            return indices.get(indexOrAlias);
-        }
-
-        Index resolveIfAbsent(DocWriteRequest<?> request) {
-            Index concreteIndex = indices.get(request.index());
-            if (concreteIndex == null) {
-                boolean includeDataStreams = request.opType() == DocWriteRequest.OpType.CREATE;
-                try {
-                    concreteIndex = indexNameExpressionResolver.concreteWriteIndex(
-                        state,
-                        request.indicesOptions(),
-                        request.indices()[0],
-                        false,
-                        includeDataStreams
-                    );
-                } catch (IndexNotFoundException e) {
-                    if (includeDataStreams == false && e.getMetadataKeys().contains(EXCLUDED_DATA_STREAMS_KEY)) {
-                        throw new IllegalArgumentException("only write ops with an op_type of create are allowed in data streams");
-                    } else {
-                        throw e;
-                    }
-                }
-                indices.put(request.index(), concreteIndex);
-            }
-            return concreteIndex;
+        IndexAbstraction resolveIfAbsent(DocWriteRequest<?> request) {
+            return indexAbstractions.computeIfAbsent(
+                request.index(),
+                key -> indexNameExpressionResolver.resolveWriteIndexAbstraction(state, request)
```

Comment:
Will this cache the index name to the index resolved for the request? I was thinking it'd have to be something like:
```
return indexAbstractions.computeIfAbsent(request.index()).resolveWriteIndexAbstraction(request);
```

Reply:
The `state` is needed for the `resolveWriteIndexAbstraction(...)` method to be able to resolve specified index expression on the request to an index, alias or data stream. We use the same cluster state instance throughout the bulk request handling here.